dev/update-dep-lock.sh

```diff
@@ -1,19 +1,61 @@
-#!/bin/sh
+#!/bin/bash
 
 # TODO: add option to update only our dependencies once conda-lock bug is fixed
 # bug url: https://github.com/conda/conda-lock/issues/652
 
+freeze_concepts=no
+
 # make sure we're in the right directory
-if [ ! -f pyproject.toml ]; then
+if [[ ! -f pyproject.toml ]]; then
     echo "this script must be run from the project root" >&2
     exit 64
 fi
 
+while [[ -n "$1" ]]; do
+    case "$1" in
+        --freeze-concepts) freeze_concepts=yes; shift;;
+        --*)
+            echo "invalid flag $1" >&2; exit 2;;
+        *)
+            echo "unexpected argument $1" >&2; exit 2;;
+    esac
+done
+
+declare -a lock_args
+lock_args=(-f pyproject.toml -f dev/environment.yml -f dev/constraints.yml)
+
+# check if we want to freeze popropx-concepts
+if [[ $freeze_concepts = yes ]]; then
+    declare -a concept_urls
+    # extract the currently-locked POPROX URL
+    concept_urls=($(yq '.package |map(select(.name == "poprox-concepts")) | map(.url) | unique() | join("\n")' conda-lock.yml))
+    if [[ ${#concept_urls[@]} -eq 1 ]]; then
+        echo "found locked poprox-concepts at $concept_urls"
+    elif [[ ${#concept_urls[@]} -eq 0 ]]; then
+        echo "ERROR: no poprox-concepts not found in lockfile" >&2
+        exit 3
+    else
+        echo "WARNING: lockfile has multiple poprox-concepts versions" >&2
+        for ver in "${concept_urls[@]}"; do
+            echo "- $ver" >&2
+        done
+        echo "using: ${concept_urls[0]}" >&2
+    fi
+
+    cat >dev/poprox-concepts-dynamic-constraint.yml <<EOF
+dependencies:
+- pip:
+    - ${concept_urls[0]}
+EOF
+    lock_args=("${lock_args[@]}" -f dev/poprox-concepts-dynamic-constraint.yml)
```

Comment:
We implement the constraint by writing an additional constraint file and including it in the `conda-lock` inputs.